playbooks/control_plane.yml

```diff
@@ -0,0 +1,18 @@
+---
+- import_playbook: init/main.yml
+
+- import_playbook: openshift-checks/private/install.yml
+
+- import_playbook: openshift-etcd/private/config.yml
+
+- import_playbook: openshift-nfs/private/config.yml
+  when: groups.oo_nfs_to_config | default([]) | count > 0
+
+- import_playbook: openshift-loadbalancer/private/config.yml
+  when: groups.oo_lb_to_config | default([]) | count > 0
+
+- import_playbook: openshift-master/private/config.yml
+
+- import_playbook: openshift-master/private/additional_config.yml
+
+- import_playbook: openshift-node/private/config.yml
```

Comment:
We could just add install_gcp step here, it only runs against masters, just make it controlled by inventory; no need to split these steps up.  In the gcp playbook above, that's really the only difference is that there's an additional play called in the middle of deploy_cluster.yml

Reply:
My concern is that long term we are going to have a very hard split here.  If we're not ready to do that it's fine, but I don't want to see coupling between "have APIs running" and "install opinionated components" for much longer.

Reply:
I don't see any reason the control plane needs to know about GCP at this point - that's a GCP specific detail, or an AWS specific detail, not something the generic playbook should be aware of.  We should be using composition, not inheritance, for many of these areas (where we are leveraging openshift-ansible as lower level playbooks)

Reply:
> My concern is that long term we are going to have a very hard split here. If we're not ready to do that it's fine, but I don't want to see coupling between "have APIs running" and "install opinionated components" for much longer.

We have done a lot of decoupling in this release.  Many people want a 'run this one command line' operation to install everything.

I suggest putting the install_gcp play here for simplicity's sake.  It's not a lot of tasks to skip if it's not normally used, and it saves you from having to include a bunch of plays in your playbook.

deploy_cluster.yml is a special, top-level play.  It utilizes re-usable plays.  I don't see value in splitting it at this time, as this gcp play is the only one that needs that to happen and we can seamlessly add the install_gcp play into deploy_cluster.

Reply:
IMO, the ideal workflow is:
1. Deploy hosts, either manually or via a play we provide
2. Prequisites and user-provided setup tasks (cloud-init script, what have you)
3. deploy_cluster.yml

Each individual cloud provider is going to have it's own set up unique steps that need to be performed; using image-based deployments is a whole other topic altogether.  I think deploy_cluster should be provider agnostic; the more we can have all providers converge on the same path, the better.

Reply:
For cloud providers (including OpenStack), the flow is going to end up looking more like:

1. Use our golden image, or customize your own, where nodes are already preconfigured
2. Deploy one or more core scale groups, using those images (that already have nodes configured)
3. Install containerized components onto the masters (not RPM based, but image based) such that the control plane comes up
4. Accept / bootstrap the master nodes and any other initial capacity provided
5. Install core components like the autoscaler and autosigner
6. Install additional components like the router, registry, etc

However, we are not going to be driving these from unified playbooks for much longer - instead, the cluster operator is going to be invoking provisioning in what is probably going to be three stages

1. provision cloud topology (load balancers, etc) and scale groups
2. provision control plane onto a set of nodes
3. install individual components

Each of those is going to be an entrypoint playbook that needs to share code with deploy_cluster - deploy_cluster is not really going to be called by anyone except people on bare metal.  I agree on bare metal the workflow you described is probably ok, except that's only 25% of the install footprint and not a priority.  We should not be optimizing for "deploy_cluster.yml" is the single entrypoint any longer.  We need to be treating the three steps as discrete, and make sure that we are reusing those pieces.  Deploy_cluster can be cloud provider agnostic, but only because it won't get called :)

Reply:
I definitely prefer not having a single playbook try to do all things.  I think it makes sense to have them be completely separate concerns.

> Each of those is going to be an entrypoint playbook that needs to share code with deploy_cluster - deploy_cluster is not really going to be called by anyone except people on bare metal. I agree on bare metal the workflow you described is probably ok, except that's only 25% of the install footprint and not a priority.

While I think many users will not be using actual bare metal hosts, many of them will be provisioning hosts manually on a variety of clouds and virt infrastructure.  I imagine some users will be quite happy with the instance/infra provisioning plays, and others are not.

I think there is a lot of value in having a simple deploy process of
1. Provision some elbs
2. Provision some hosts
3. Provision openshift

We had discussed previously making all the nodes 'bootstrap-able'.  I think we should try to move that direction so we have as common as possible install experience across all deployment types.